cmd/bosun/sched/sched.go

```diff
@@ -540,45 +540,55 @@ func init() {
 		"The running count of actions performed by individual users (Closed alert, Acknowledged alert, etc).")
 }
 
-func (s *Schedule) ActionByAlertKey(user, message string, t models.ActionType, ak models.AlertKey) error {
+func (s *Schedule) ActionByAlertKey(user, message string, t models.ActionType, at *time.Time, ak models.AlertKey) error {
 	st, err := s.DataAccess.State().GetLatestIncident(ak)
 	if err != nil {
 		return err
 	}
 	if st == nil {
 		return fmt.Errorf("no such alert key: %v", ak)
 	}
-	_, err = s.action(user, message, t, st)
+	_, err = s.action(user, message, t, at, st)
 	return err
 }
 
-func (s *Schedule) ActionByIncidentId(user, message string, t models.ActionType, id int64) (models.AlertKey, error) {
+func (s *Schedule) ActionByIncidentId(user, message string, t models.ActionType, at *time.Time, id int64) (models.AlertKey, error) {
 	st, err := s.DataAccess.State().GetIncidentState(id)
 	if err != nil {
 		return "", err
 	}
 	if st == nil {
 		return "", fmt.Errorf("no incident with id: %v", id)
 	}
-	return s.action(user, message, t, st)
+	return s.action(user, message, t, at, st)
 }
 
-func (s *Schedule) action(user, message string, t models.ActionType, st *models.IncidentState) (ak models.AlertKey, e error) {
+func (s *Schedule) action(user, message string, t models.ActionType, at *time.Time, st *models.IncidentState) (ak models.AlertKey, e error) {
 	if err := collect.Add("actions", opentsdb.TagSet{"user": user, "alert": st.AlertKey.Name(), "type": t.String()}, 1); err != nil {
 		slog.Errorln(err)
 	}
+	isUnknown := st.LastAbnormalStatus == models.StUnknown
+	timestamp := utcNow()
+	action := models.Action{
+		Message: message,
+		Time:    timestamp,
+		Type:    t,
+		User:    user,
+	}
 	defer func() {
 		if e == nil {
 			if err := collect.Add("actions", opentsdb.TagSet{"user": user, "alert": st.AlertKey.Name(), "type": t.String()}, 1); err != nil {
 				slog.Errorln(err)
 			}
-			if err := s.DataAccess.Notifications().ClearNotifications(st.AlertKey); err != nil {
-				e = err
+			switch action.Type {
```

Comment:
The point of defer was to make sure the notifications get cleared no matter what. Perhaps the alternative of clearing in each case is looking better now? A switch with one case should be an if, regardless.